### PR TITLE
Adds EXTRACT function for extracting fields from DATETIME

### PR DESCRIPTION
**Pull request by jianqiao**
This PR adds parser and resolver support for the `EXTRACT` function.

Since the backend `DateExtractOperation` and `DateExtractUncheckedOperator` are already there. So the `EXTRACT` function feature is complete. SQL tests are also added.

### CODE REVIEW

File: query_optimizer/tests/execution_generator/Select.test [hunk 1]
```diff
@@ -632,6 +632,48 @@ ORDER BY group_col1,
 +--------------------+--------------------+--------------------+
 ==
 
+
+CREATE TABLE dates (value DATETIME);
```
Comment (zuyu) [inline query_optimizer/tests/execution_generator/Select.test]: If we change one test file in `query_optimizer`, we need to update other three test files under different directories. In this case, they are

 * `resolver`
 * `logical_generator`
 * `physical_generator`

File: parser/ParseBasicExpressions.hpp [hunk 2]
```diff
@@ -448,6 +448,69 @@ class ParseFunctionCall : public ParseExpression {
...
+   *
+   * @param line_number The line number of the token "extract" in the statement.
+   * @param column_number The column number of the token "extract in the statement.
+   * @param extract_field The field to extract.
+   * @param source_expression The expression to extract a field from.
+   */
+  ParseExtractFunction(const int line_number,
+                       const int column_number,
+                       ParseString *extract_field,
+                       ParseExpression *date_expression)
```
Comment (zuyu) [inline parser/ParseBasicExpressions.hpp]: Could we mark two pointers above `const`?

File: parser/ParseBasicExpressions.hpp [hunk 3]
```diff
@@ -448,6 +448,69 @@ class ParseFunctionCall : public ParseExpression {
...
+      std::vector<std::string> *inline_field_names,
+      std::vector<std::string> *inline_field_values,
+      std::vector<std::string> *non_container_child_field_names,
+      std::vector<const ParseTreeNode*> *non_container_child_fields,
+      std::vector<std::string> *container_child_field_names,
+      std::vector<std::vector<const ParseTreeNode*>> *container_child_fields) const override;
+
+ private:
+  std::unique_ptr<ParseString> extract_field_;
+  std::unique_ptr<ParseExpression> date_expression_;
```
Comment (zuyu) [inline parser/ParseBasicExpressions.hpp]: Mark two pointers `const`.

File: parser/ParseExpression.hpp [hunk 4]
```diff
@@ -37,10 +39,11 @@ class ParseExpression : public ParseTreeNode {
   enum ExpressionType {
     kAttribute,
     kBinaryExpression,
+    kExtract,
     kFunctionCall,
     kScalarLiteral,
     kSubqueryExpression,
-    kUnaryExpression,
+    kUnaryExpression
```
Comment (zuyu) [inline parser/ParseExpression.hpp]: Reset this change. `c++ 11` allows it.

**Comment by zuyu** [comment]
@jianqiao In the title, replace `operator` with `function`.

File: query_optimizer/tests/execution_generator/Select.test (hunk 1 above)
Comment (jianqiao) [inline query_optimizer/tests/execution_generator/Select.test]: I updated a test for `resolver`. For `logical_generator` and `physical_generator`, I think the purpose  is to check query plans. But `EXTRACT` is an expression which is not changed during logical planning and physical planning. So the resolver test should be sufficient.

File: parser/ParseBasicExpressions.hpp (hunk 2 above)
Comment (jianqiao) [inline parser/ParseBasicExpressions.hpp]: We can do that but all the other `ParseTreeNode` subclasses did not use `const` for `ParseTreeNode` objects. Perhaps a separate PR to update all if that is a concern?

File: parser/ParseBasicExpressions.hpp (hunk 3 above)
Comment (jianqiao) [inline parser/ParseBasicExpressions.hpp]: Same concern as above.

File: parser/ParseExpression.hpp (hunk 4 above)
Comment (jianqiao) [inline parser/ParseExpression.hpp]: Updated.

**Comment by pateljm** [comment]
LTGM. Merging. I see the Travis error, but that is my bad as the line is over 120 characters and I'll fix it next in a separate PR. 

Thanks @jianqiao.

**Pull request merged**